### PR TITLE
Expose DocumentationCommentId as public API

### PR DESCRIPTION
**Pull request by mattwar**
This API is useful for consuming documentation comment ID's that appear in documentation files generated by the compiler. It can also be used to produce them given a symbol.

This is a proposal that we expose this API.

@Pilchie, @srivatsn  please review

### CODE REVIEW

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 1]
```diff
@@ -69,19 +79,45 @@ public static string CreateReferenceId(ISymbol symbol)
...
+            }
+
+            if (compilation == null)
+            {
+                throw new ArgumentNullException(nameof(compilation));
+            }
+
+            var results = new List<ISymbol>();
+            TryGetSymbolsForDeclarationId(id, compilation, results);
             return results;
```
Comment (Pilchie) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs]: Should we make an actual safe copy here that can't be casted away?

Comment (mattwar) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs]: probably.

**Comment by Pilchie** [comment]
:+1:

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 2]
```diff
@@ -69,19 +79,52 @@ public static string CreateReferenceId(ISymbol symbol)
...
         }
 
         /// <summary>
-        /// Try get all declaration symbols that match the declaration id string
+        /// Try to get all the declaration symbols that match the declaration id string.
+        /// Returns true if atleast one symbol matches.
         /// </summary>
-        public static bool TryGetSymbolsForDeclarationId(string id, Compilation compilation, out List<ISymbol> results)
+        /// <remarks>This API is available for possible of reuse of results lists.</remarks>
+        public static bool TryGetSymbolsForDeclarationId(string id, Compilation compilation, List<ISymbol> results)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs]: :bulb: I'm not sure I'm ready for this method to be in the public API, so I would consider making it `internal`.

I believe `GetSymbolsForDeclarationId` covers the essential external use cases already.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 3]
```diff
@@ -105,16 +158,58 @@ public static ISymbol GetFirstSymbolForDeclarationId(string id, Compilation comp
...
+                s_symbolListPool.ClearAndFree(results);
+            }
+        }
+
+        /// <summary>
+        /// Try to get all symbols that match the reference id string.
+        /// Returns true if atleast one symbol matches.
+        /// </summary>
+        /// <remarks>This API is available for possible of reuse of results lists.</remarks>
+        public static bool TryGetSymbolsForReferenceId(string id, Compilation compilation, List<ISymbol> results)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs]: :bulb: My suggestion here is the same as above (consider making this method internal).

**Pull request merged**